pgx-pg-sys/src/lib.rs

```diff
@@ -160,29 +162,38 @@ pub use internal::pg13::*;
 #[cfg(feature = "pg14")]
 pub use internal::pg14::*;
 
-/// A trait applied to all of Postgres' `pg_sys::Node` types and its subtypes
+/// A trait applied to all Postgres `pg_sys::Node` types and subtypes
 pub trait PgNode {
-    type NodeType;
+    // TODO(0.6.0): seal this trait or take this trait private?
 
-    /// Represent this node as a mutable pointer of its type
+    /// Format this node
     #[inline]
-    fn as_node_ptr(&self) -> *mut Self::NodeType {
-        self as *const _ as *mut Self::NodeType
+    fn display_node(&self) -> String {
+        // SAFETY: The trait is pub but this impl is private, and
+        // this is only implemented for things known to be "Nodes"
+        unsafe { display_node_impl(NonNull::from(self).cast()) }
     }
 }
 
 /// implementation function for `impl Display for $NodeType`
-pub(crate) fn node_to_string_for_display(node: *mut crate::Node) -> String {
+///
+/// # Safety
+/// Don't use this on anything that doesn't impl PgNode, or the type may be off
+#[warn(unsafe_op_in_unsafe_fn)]
+pub(crate) unsafe fn display_node_impl(node: NonNull<crate::Node>) -> String {
```

Comment:
The previous `fn node_to_string_for_display` which this function exploded from the chest of was actually `unsafe`, as a null pointer input was not handled. This is only true in a trivial sense as it was also private, but it's worth noting that e.g. making it truly `pub` would have been unsound.